### PR TITLE
Support Pre-GA Python Versions

### PR DESCRIPTION
**Pull request by scbedd**
Official policy for ADO images is to NOT ship the pre-GA bits. 

No problem, we'll lean on the infrastructure created by the actions virtual-environment crew and support it with some custom templating.

This PR essentially creates a shim between devops naming and gh-actions installer release naming.

### CODE REVIEW

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml [hunk 1]
```diff
@@ -39,6 +39,11 @@ parameters:
       PythonVersion: 'pypy3'
       CoverageArg: '--disablecov'
       RunForPR: false
+    Linux_Python39rc2:
+      OSVmImage: 'ubuntu-18.04'
+      PythonVersion: '3.9.0-rc.2'
```
Comment (zooba) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:44]: If you specify `'3.9.0-dev'` then it should get the latest prerelease version available (which won't matter so much now, but it should automatically roll over to the final release when it's available)

Comment (scbedd) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:44]: Hmm. I think I am missing something, I don't believe this is the case. The only pre-cached versions of python are GA up to 3.8.5, and the `UsePythonVersion` task does not support download :( 

I am adding download, followed by installing from the actions manifest if possible.

File: eng/pipelines/templates/steps/wrapped-use-python.yml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    inputs:
+      versionSpec: 3.8
+
+  - pwsh: |
+      pip install packaging
+    displayName: Prep Environment
+
+  # select the appropriate version from manifest if present
+  - task: PythonScript@0
+    displayName: 'Check Version Spec Against Known Versions'
```
Comment (scbedd) [inline eng/pipelines/templates/steps/wrapped-use-python.yml:17]: This name needs to be adjusted to `Install Python Version`. I'll do so next time I'm at a machine with git.

File: eng/pipelines/templates/steps/wrapped-use-python.yml [hunk 3]
```diff
@@ -0,0 +1,26 @@
+parameters:
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/wrapped-use-python.yml:1]: nit: I'd call this template just `use-python-version.yml`

File: eng/pipelines/templates/steps/wrapped-use-python.yml [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  - pwsh: |
+      pip install packaging
+    displayName: Prep Environment
+
+  # select the appropriate version from manifest if present
+  - task: PythonScript@0
+    displayName: 'Check Version Spec Against Known Versions'
+    inputs:
+     scriptPath: 'scripts/devops_tasks/resolve_installer_location.py'
+     arguments: '${{ parameters.versionSpec }} --installer_folder="../_pyinstaller'
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/wrapped-use-python.yml:20]: Is it safe to assume that `../_pyinstaller` is the same place the UsePythonVersion task will look for the installed versions? Looks like it uses Agent.ToolsDirectory based on the docs at https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/tool/use-python-version?view=azure-devops

File: scripts/devops_tasks/resolve_installer_location.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    parser.add_argument(
+        "--installer_folder",
+        dest="installer_folder",
+        help=(
+            "The folder where the found installer will be extracted into and run from."
+        ),
+    )
+
+    args = parser.parse_args()
+    max_precached_version = Version("3.8.6")
```
Comment (weshaggard) [inline scripts/devops_tasks/resolve_installer_location.py:137]: Do we have any plans to update this? Does it need to be updated?

File: eng/pipelines/templates/steps/wrapped-use-python.yml (hunk 3 above)
Comment (scbedd) [reply eng/pipelines/templates/steps/wrapped-use-python.yml:1]: 👍

File: eng/pipelines/templates/steps/wrapped-use-python.yml (hunk 4 above)
Comment (scbedd) [reply eng/pipelines/templates/steps/wrapped-use-python.yml:20]: No! The installer actually unzips and installs into the precached tools directory! I'm just saving the installer to the disk so I can unzip it and run it. I owe this PR a rename run after I merged the script that installs into this script.

File: scripts/devops_tasks/resolve_installer_location.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+import os
+import zipfile
+import tarfile
+import time
+
+from packaging.version import Version
+from packaging.version import parse
+from packaging.version import InvalidVersion
+
+import pdb
```
Comment (scbedd) [inline scripts/devops_tasks/resolve_installer_location.py:17]: Will cut.

File: scripts/devops_tasks/resolve_installer_location.py (hunk 5 above)
Comment (scbedd) [reply scripts/devops_tasks/resolve_installer_location.py:137]: Yes. Will need to update this once the next major version of python comes out. Once a year. Given that we'll be on gh-actions eventually (and we'll toss this entire thing), I'm betting we'll do that exactly once.

**Comment by scbedd** [comment]
@yunhaoling python 3.9 isn't pre-cached still. we can take this PR, which will unblock your use of python 3.9 in your service bus PR.

**Review by lmazuel — APPROVED** [review]

**Review by weshaggard — APPROVED** [review]

**Comment by yunhaoling** [comment]
@scbedd thanks for helping releasing uamqp, I re-run the eventhub CI in your PR and it's happy now.
Do you consider taking Brandon's change to the install python version script into this PR or not?

**Comment by scbedd** [comment]
@yunhaoling that's a great point. I'll mirror it.

**Pull request merged**